### DISCUSSION TITLE
Generate Thesis pdf with pandoc

### DISCUSSION THREAD

**Discussion by coppertank** [cagix/pandoc-thesis#62] [Q&A]
Hi everyone, I apologize in advance for my noobie question but I don't have much experience with pandoc. 
Reading the installation section for creating the thesis in PDF, there are two ways: one with docker and the other with pandoc. 

Point 4, specifically, says: "Either set an environment variable `PANDOC` pointing to the location of your pandoc installation or change the line `PANDOC      ?= docker run --rm -v $(WORKDIR):/pandoc pandoc-thesis pandoc` in the `Makefile` to reflect your settings". 

However, I'm not sure how to interpret this point. 
Suppose I want to produce the PDF of the thesis with pandoc (Eisvogel template), once I have downloaded the zip containing all the files, what would be the steps and the code to generate it with pandoc? 

Thanks in advance

**Comment by cagix** [comment]
hmm, so are you really going to take the _scenic route_? in this case:

1. determine the path to the pandoc binary: for linux/macos you could use e.g. `which pandoc` 
2. this returns the path to pandoc, e.g. `/usr/local/bin/pandoc`
3. use this string to replace the right-hand side of the line in the makefile, i.e. `PANDOC      ?= /usr/local/bin/pandoc`. 
 
please remember that you also need a working latex installation with all the packages mentioned.

**Reply by coppertank** [reply]
Thank you for the reply, but I have some more questions. 
Once I have replaced the line in the Makefile, the only thing to do is open a terminal and do the command `make simple`?

Because doing it like this, i get the following error:

C:/Users/caumo/AppData/Local/Pandoc --wrap=preserve --template=titlepage.tex --metadata-file=md/metadata.yaml -o __titlepage.filled.tex titlepage.tex
process_begin: CreateProcess(NULL, C:/Users/caumo/AppData/Local/Pandoc --wrap=preserve --template=titlepage.tex --metadata-file=md/metadata.yaml -o __titlepage.filled.tex titlepage.tex, ...) failed.
make (e=5): Accesso negato.
make: *** [Makefile:196: __titlepage.filled.tex] Error 5


Am I missing something?

**Reply by cagix** [reply]
this should be the only change required, i.e. after that change `make simple` should produce a pdf.

i do not have access to a windows machine. what happens, if you execute the failed command line directly in a terminal, in this case `C:/Users/caumo/AppData/Local/Pandoc --wrap=preserve --template=titlepage.tex --metadata-file=md/metadata.yaml -o __titlepage.filled.tex titlepage.tex`? 

are the indented lines in the makefile still indented by tabs? some editors convert tabs to spaces, which would be fatal in this case?

**Reply by cagix** [reply]
@coppertank i just googled "makefile error 5". it seems you are not allowed to run commands in a terminal on your windows machine: 
- https://answers.microsoft.com/en-us/windows/forum/all/how-to-fix-error-5-access-is-denied-in-windows-10/8a4069ea-4b0e-4517-9d7a-62c44b4d6e7d
- https://www.thewindowsclub.com/system-error-5-access-is-denied-in-windows ...

**Reply by coppertank** [reply]
Thank you so much, I will do some testing. 
Just to be sure, opening the makefile in VScode to change the variable, how do i recognize the tabs instead of the spaces? Where are they located?

**Reply by cagix** [reply]
> Just to be sure, opening the makefile in VScode to change the variable, how do i recognize the tabs instead of the spaces? Where are they located?

each indented line should be indented by a single tab. while there certainly is some setting in vsc to show tabs vs. spaces, you could just place the cursor at the beginning of an indented line und move it to the right using the arrow-keys on your keyboard. if there is a tab in the beginning of the line, a single touch on the right-arrow-key will "jump" the cursor to the first word. if there are spaces instead, the cursor will advance just one blank and will be somewhere between the beginning of the line and the first word.

**Reply by coppertank** [reply]
Vscode sees the tabs correctly, i run the command prompt with administrator rights and tried to run the command several times but nothing (I also added the pandoc folder to the environments variable).

However, changing the Pandoc path in `C:/Users/caumo/AppData/Local/Pandoc` gives me the same error 5: access denied, while pointing at the exact file pandoc.exe (`C:/Users/caumo/AppData/Local/Pandoc/pandoc.exe`) it gives me another error:

```
C:/Users/caumo/AppData/Local/Pandoc/pandoc --wrap=preserve --template=titlepage.tex --metadata-file=md/metadata.yaml -o __titlepage.filled.tex titlepage.tex
C:/Users/caumo/AppData/Local/Pandoc/pandoc --wrap=preserve --template=frontmatter.tex --metadata-file=md/metadata.yaml -o __frontmatter.filled.tex frontmatter.tex
C:/Users/caumo/AppData/Local/Pandoc/pandoc --wrap=preserve --template=backmatter.tex --metadata-file=md/metadata.yaml -o __backmatter.filled.tex backmatter.tex
C:/Users/caumo/AppData/Local/Pandoc/pandoc -f markdown --pdf-engine=pdflatex --standalone -M lang=de-DE --metadata-file=md/metadata.yaml --include-in-header=__titlepage.filled.tex --include-before-body=__frontmatter.filled.tex --include-after-body=__backmatter.filled.tex --citeproc -M bibliography=references.bib -M link-citations=true --listings -V documentclass=scrbook -V papersize=a4 -V fontsize=11pt -V classoption:open=right -V classoption:twoside=true -V classoption:cleardoublepage=empty -V classoption:clearpage=empty -V geometry:top=30mm -V geometry:left=25mm -V geometry:bottom=30mm -V geometry:width=150mm -V geometry:bindingoffset=6mm --toc --toc-depth=3 --number-sections -V colorlinks=true -V book=true -V titlepage=true -V toc-own-page=true -o thesis.pdf md/introduction.md md/relatedwork.md md/concept.md md/realisation.md md/conclusion.md references.md md/appendix.md
pandoc: C:\Users\caumo\Desktop\pandoc-thesis-master\__titlepage.filled.tex __frontmatter.filled.tex __backmatter.filled.tex\tex2pdf.-17ccb46edec0caa0: CreateDirectory "\\\\?\\C:\\Users\\caumo\\Desktop\\pandoc-thesis-master\\__titlepage.filled.tex __frontmatter.filled.tex __backmatter.filled.tex\\tex2pdf.-17ccb46edec0caa0": does not exist (Impossibile trovare il percorso specificato.)
make: *** [Makefile:191: thesis.pdf] Error 1
```

In this latter case, i can see in the folder where i have all the files that pandoc could build the temporary files `__backmatter.filled.tex`, `__frontmatter.filled.tex` and `__titlepage.filled.tex`, while in the first case no. However, i think the output suggests that a specific file couldn't be found, I don't know what to do

**Reply by cagix** [reply]
> However, changing the Pandoc path in `C:/Users/caumo/AppData/Local/Pandoc` gives me the same error 5: access denied, while pointing at the exact file pandoc.exe (`C:/Users/caumo/AppData/Local/Pandoc/pandoc.exe`) 

ah. on a unix box `/usr/local/bin/pandoc` is the executable, so i assumed `C:/Users/caumo/AppData/Local/Pandoc` would be the pandoc binary on your machine. so in this case you would just need to put `C:/Users/caumo/AppData/Local/Pandoc/pandoc.exe` into the makefile :)




> C:/Users/caumo/AppData/Local/Pandoc/pandoc --wrap=preserve --template=titlepage.tex --metadata-file=md/metadata.yaml -o __titlepage.filled.tex titlepage.tex
> 
> C:/Users/caumo/AppData/Local/Pandoc/pandoc --wrap=preserve --template=frontmatter.tex --metadata-file=md/metadata.yaml -o __frontmatter.filled.tex frontmatter.tex
> 
> C:/Users/caumo/AppData/Local/Pandoc/pandoc --wrap=preserve --template=backmatter.tex --metadata-file=md/metadata.yaml -o __backmatter.filled.tex backmatter.tex
> 
> C:/Users/caumo/AppData/Local/Pandoc/pandoc -f markdown --pdf-engine=pdflatex --standalone -M lang=de-DE --metadata-file=md/metadata.yaml --include-in-header=__titlepage.filled.tex --include-before-body=__frontmatter.filled.tex --include-after-body=__backmatter.filled.tex --citeproc -M bibliography=references.bib -M link-citations=true --listings -V documentclass=scrbook -V papersize=a4 -V fontsize=11pt -V classoption:open=right -V classoption:twoside=true -V classoption:cleardoublepage=empty -V classoption:clearpage=empty -V geometry:top=30mm -V geometry:left=25mm -V geometry:bottom=30mm -V geometry:width=150mm -V geometry:bindingoffset=6mm --toc --toc-depth=3 --number-sections -V colorlinks=true -V book=true -V titlepage=true -V toc-own-page=true -o thesis.pdf md/introduction.md md/relatedwork.md md/concept.md md/realisation.md md/conclusion.md references.md md/appendix.md
> 
> pandoc: C:\Users\caumo\Desktop\pandoc-thesis-master\__titlepage.filled.tex __frontmatter.filled.tex __backmatter.filled.tex\tex2pdf.-17ccb46edec0caa0: CreateDirectory "\\\\?\\C:\\Users\\caumo\\Desktop\\pandoc-thesis-master\\__titlepage.filled.tex __frontmatter.filled.tex __backmatter.filled.tex\\tex2pdf.-17ccb46edec0caa0": does not exist (Impossibile trovare il percorso specificato.)


so, all the temporary files (starting with "`__`" are created, but the last step goes wild? 

i noticed in the output that the path to the first auxiliary file is repeated using the full path, but the others are apparently not being recognised is individual files but concatenated to form a sort of massive filename? somehow pandoc seems to not recognise the options?

you are using "/" as path separator. iirc in windows your are required to use "\" instead? could you copy the last cmd to your terminal but replace every "/" with "\"? 

sadly i can't test this as i have no access to a windows machine.

**Reply by coppertank** [reply]
Ok I was able to knit in pdf the simple example, now i will try the other methods and I will report if there are any other problems.

I was able to generate the pdf in this way:
- first, after changing the line in the makefile with my pandoc directory (pointing at the exe file) i run `make simple` to generate the temporary filled tex files. The procedure still give me error, though.
- i then copied the whole code before the error message (from `C:/Users/caumo/AppData/Local/Pandoc/pandoc.exe -f markdown`... to `references.md md/appendix.md`) and pasted it another time in the command prompt, changing the line `C:/Users/caumo/AppData/Local/Pandoc/pandoc.exe` with only `pandoc`.

This was the output:
```
pdflatex: security risk: running with elevated privileges
pdflatex: security risk: running with elevated privileges
[WARNING] [makePDF] LaTeX Warning: Unused global option(s): [clearpage].
```

In the folder i found then the `thesis.pdf` file, the temporary files were not automatically deleted but for me this was not a big issue

Thank you so much for the patience

Edit: the same procedure works also changing the right side of the Makefile with only `pandoc`. However, i need all the times to run the procedure twice (first `make simple` and then to copy/paste the last part of the code), so i think there is some trouble with automated actions

**Reply by cagix** [reply]
glad it worked! in general you can either use the complete path (including the executable) or just the name itself. the latter will only work, if the executable is on your search path. so in your case it should definitely work using "c:\\...\pandoc.exe" or just "pandoc" (assuming the path is set up correctly).

i'm planning on using pandoc default files. this would make things a bit easier ...

**Reply by cagix** [reply]
> the temporary files were not automatically deleted

@coppertank `make clean` should remove all temporary artefacts :)
